### PR TITLE
Go modules ignore conditions

### PR DESCRIPTION
**Pull request by mctofu**
This brings in support for ignored versions for the go modules ecosystem in dependabot-core. It also extracts a `LatestVersionFinder` class which is common in the other ecosystems.

This was an alternate approach to https://github.com/dependabot/dependabot-core/pull/3630. Instead of migrating to use `go list -m -versions` we stuck to using our existing `updatechecker` helper but modified it to work more like `go list` with these changes:
- Returns all allowed versions instead of only the latest version
- Filters by major version and go mod excluded versions (retracted versions need extra work to support)
- Removed pre-release/psuedo version filtering which is moved to the ruby side

With those changes in place applying the ignored version filter was fairly simple: https://github.com/dependabot/dependabot-core/pull/3631/commits/2bdbabce2903ae59213659f17064f763397bfd92

### CODE REVIEW

File: go_modules/lib/dependabot/go_modules/update_checker/latest_version_finder.rb [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+        def filter_prerelease_versions(versions_array)
+          return versions_array if wants_prerelease?
+
+          versions_array.reject(&:prerelease?)
+        end
+
+        def filter_lower_versions(versions_array)
+          versions_array.
+            select { |version| version > Gem::Version.new(dependency.version) }
```
Comment (mctofu) [inline go_modules/lib/dependabot/go_modules/update_checker/latest_version_finder.rb:111]:
```suggestion
            select { |version| version > version_class.new(dependency.version) }
```

**Review by thepwagner — APPROVED** [review]
LGTM - I'm excited to ignore some versions.
Most of my time on this review was spent contemplating [the path not taken](https://github.com/dependabot/dependabot-core/pull/3630) - thanks for the clear and documented reasoning there!

File: go_modules/lib/dependabot/go_modules/update_checker.rb [hunk 2]
```diff
@@ -33,7 +25,12 @@ def latest_resolvable_version
         end
 
         @latest_resolvable_version ||=
-          version_class.new(find_latest_resolvable_version.gsub(/^v/, ""))
+          LatestVersionFinder.new(
+            dependency: dependency,
+            dependency_files: dependency_files,
+            credentials: credentials,
+            ignored_versions: ignored_versions,
```
Comment (feelepxyz) [inline go_modules/lib/dependabot/go_modules/update_checker.rb:32]:
```suggestion
            ignored_versions: ignored_versions,
            raise_on_ignored: raise_on_ignored
```

File: go_modules/lib/dependabot/go_modules/update_checker/latest_version_finder.rb [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+          candidate_versions = filter_lower_versions(candidate_versions)
+          candidate_versions = filter_ignored_versions(candidate_versions)
+
+          candidate_versions.max
+        end
+
+        def available_versions
+          SharedHelpers.in_a_temporary_directory do
+            SharedHelpers.with_git_configured(credentials: credentials) do
+              File.write("go.mod", go_mod.content)
```
[the remark this answers is not in the document]
Comment (mctofu) [reply go_modules/lib/dependabot/go_modules/update_checker/latest_version_finder.rb:55]: Hm, good call! It's probably not needed but I'm hesitant to change it as part of this PR. This was just moved over from the original implemenation: https://github.com/dependabot/dependabot-core/blob/main/go_modules/lib/dependabot/go_modules/update_checker.rb#L60

File: go_modules/lib/dependabot/go_modules/update_checker/latest_version_finder.rb [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        def latest_version
+          @latest_version ||= fetch_latest_version
+        end
+
+        private
+
+        attr_reader :dependency, :dependency_files, :credentials, :ignored_versions
+
+        def fetch_latest_version
+          pseudo_version_regex = /\b\d{14}-[0-9a-f]{12}$/
```
[the remark this answers is not in the document]
Comment (mctofu) [reply go_modules/lib/dependabot/go_modules/update_checker/latest_version_finder.rb]: Done in https://github.com/dependabot/dependabot-core/pull/3631/commits/9e402ae128ed4d5688ba37a77385362cc97f1ed6

**Pull request merged**